Implementation/68812 create basic portfolio view for all portfolios

EinLama opened the pull request:
# Ticket
https://community.openproject.org/wp/68812

# What are you trying to accomplish?
Introduce the first implementation chunk for the portfolio module that provides a basic feature set and is usable as-is. Includes:

* menu items in the global menu (grid button top left) as well as in the global sidebar menu. Only visible if the user has the `add_portfolios` permission or has the `view_project` permission on any active Portfolio. Additionally, the feature flag is checked.
* Portfolios controller - a copy of the projects controller with a reduced feature set. Includes authorization. Only offers the `index` action so far. Note that the projects controller is still handling the `new` and `create` actions for portfolios.
* An index view containing some new components for displaying Portfolios.
  * Reuses the Project filter component. For now, all filters are identical to regular project filters. We might change this in the future.
  * The index view relies on static queries to fetch the Portfolios. Due to ease of implementation, I added new static queries to `ProjectQueries` and applied a `.portfolio` scope to it on the result set prior to displaying. This was the least intrusive way I found for fetching the desired portfolios.

## Screenshots
<img width="1099" height="606" alt="image" src="https://github.com/user-attachments/assets/71674f41-c7c8-4efe-ab0f-83cac4c6c14c" />

# Merge checklist

- [x] Added/updated tests
- [ ] Added/updated documentation in Lookbook (patterns, previews, etc)
- [ ] Tested major browsers (Chrome, Firefox, Edge, ...)

EinLama committed:
[#68812] WIP

EinLama committed:
[#68812] WIP - use a filter to show portfolios

EinLama committed:
[#68812] WIP - with scope 'portfolio'

EinLama committed:
[#68812] Revert "[#68812] WIP - use a filter to show portfolios"

This reverts commit 76b4c9d007565add991e41da369410c7524bcd69.

EinLama committed:
[#68812] Add menu entries for portfolios

EinLama committed:
[#68812] Portfolio preview links to portfolio

EinLama committed:
[#68812] Rough UI mockup

EinLama committed:
[#68812] Show relative time for updated at

EinLama committed:
[#68812] Portfolio list styling

EinLama committed:
[#68812] Collect child programs and projects

EinLama committed:
[#68812] Show favorited status correctly

EinLama committed:
[#68812] Use turbo for page update

EinLama committed:
[#68812] Add TODOs

EinLama committed:
[#68812] Portfolio side menu and static portfolio queries

EinLama committed:
[#68812] Reuse exisitng no results string

EinLama committed:
[#68812] Remove unneeded methods

EinLama committed:
[#68812] Add TODOs for future features

EinLama committed:
[#68812] Check portfolio feature flag

EinLama committed:
[#68812] Remove unneeded columns param

EinLama committed:
[#68812] Remove export TODO

This will not be implemented for a while.

EinLama committed:
[#68812] Makeshift portfolio permission check

EinLama committed:
[#68812] Clean up

Remove unneeded methods, smaller refactors

EinLama committed:
[#68812] false

EinLama committed:
[#68812] Portfolio permission check within controller

EinLama committed:
[#68812] Use existing `add_portfolios` permission

EinLama committed:
[#68812] render_403

EinLama committed:
[#68812] Rename portfolio components

The previous names felt wrong and a bit too generic when compared to the
other existing components in our application. Tried to make the names
sound more like the existing ones.

mv {List,Index}Component
mv {Preview,Details}Component

EinLama committed:
[#68812] Add missing projects controller spec

There was a shared example, but it was never used. Why does our static
code checking miss this?

EinLama committed:
[#68812] PortfoliosController spec

EinLama committed:
[#68812] Portfolio component specs

EinLama committed:
[#68812] Portfolio feature spec

EinLama committed:
[#68812] Remove unnecessary rubocop:disable

ulferts commented on app/components/portfolios/details_component.rb:
```diff
@@ -26,33 +26,54 @@
-    redirect_to project_settings_creation_wizard_path(@project, tab: params[:tab]), status: :see_other
-  end
+    def all_subprojects
+      all_descendants.filter { it.workspace_type == "project" }
```
This can also be written as `it.project?` since an enum is used in the model.

EinLama replied:
Good point. I ended up using `all_descendants.project` to further simplify it.

ulferts commented on app/components/portfolios/details_component.rb:
```diff
@@ -26,33 +26,54 @@
-  def check_feature_flag
-    unless OpenProject::FeatureDecisions.project_initiation_active?
-      render_404
+      @descendants
```
I don't get, why this is necessary here. `.descendants` will give you all descendants, not only the children. I therefore think that:
```
@descendants ||= portfolio.descendants.visible
```
is sufficient. Please note that I added the `visible` scope here as only those descendants the user is allowed to see should be returned.

EinLama replied:
This is awesome, thanks!

My bad. We have various means of accessing child projects. `#children`, `#hiearchy`, `#descendants`. I thought I looked at _all_ of these when comparing. But I missed that `descendants` already returns the nested children. I'm glad I can scrap this ugly method.

ulferts commented on app/components/portfolios/details_component.rb:
```diff
@@ -26,33 +26,54 @@
-    unless OpenProject::FeatureDecisions.project_initiation_active?
-      render_404
+      @descendants
     end
```
This is really the curse of components here. With the way this is written, every component for a portfolio will query for favorited and descendants. `favorited_project_ids` would run into the query cache so it would at least not hit the database. Given that this is just ids, the overhead is probably not very high (not as high as for AR objects at least).

But if the projects/portfolios would already loaded with the `with_favorited_by_user`, then one could simply call `portfolio.favorited?` here. This admittedly has the drawback of leading to non obviously coupled code. I.e. if the `with_favorited_by_user` is removed, the call to `favorited?` would run into a `NoMethodError`. But it would at least run into that error which at least then would hint at the connection. I currently would therefore prefer this to happen.

About descendants, this has the problem of not being eager-loadable, given that this is not modelled using associations. So I don't have a good solution to offer here. We could of course bulk load them somewhere but getting them into the component seems out of scope currently. It is really unfortunate that all the tree/graph libraries don't work with associations to easily transport the results of their queries.

EinLama replied:
Great catch, thanks! Preloading the favorite status now as suggested.

Agreed, not being able to preload the descendants easily is annoying. Fixing this is too expensive at the moment. Let's see first how much pain it will cause.

ulferts commented on config/routes.rb:
```diff
@@ -494,6 +494,12 @@
+    resource :menu, only: %i[show]
+  end
+
+  resources :portfolios, only: %i[index]
```
This overwrites the `resources :portfolios` statement above leading to calls to `portfolios/new` no longer carrying the workspace type with them. This then leads to projects being created instead of portfolios.

EinLama replied:
Well spotted. Had to fiddle around a bit, fixed now.

ulferts commented on app/components/portfolios/details_component.html.erb:
```diff
@@ -0,0 +1,95 @@
+    end
+
+    flex.with_row(mb: 2) do
+      @portfolio.description
```
This would need to be formatted, which changing this to

```
format_text(@portfolio, :description)
```

will do. 

But this will lead to new problems when one goes crazy and adds a lot of text to the description or just uses macros and other niceties:

<img width="2028" height="1101" alt="Image" src="https://github.com/user-attachments/assets/2c47ddcd-7336-4dec-8c0d-c0e94ae0bf5a" />

It breaks the page.

Not sure what to do on this. Some limit on what is possible, similar to what we do for the projects table might be an option but that would strip out any html which is also not what we want to do I recon. 

I'll go on and ask for opinions  in the work package.

EinLama replied:
Created https://community.openproject.org/wp/69115 to fix this issue separately.

ulferts commented on app/components/portfolios/index_sub_header_component.html.erb:
```diff
@@ -0,0 +1,37 @@
+          data: { "test-selector": "portfolio-new-button" }
+        ) do
+          t(:label_portfolio)
+        end
```
This is lacking the check for the `add_portfolios` permission.

ulferts commented on app/controllers/portfolios_controller.rb:
```diff
@@ -0,0 +1,99 @@
+
+  def authorize_portfolio_access
+    render_403 unless User.current.allowed_globally?(:add_portfolios) ||
+                       Project.active.portfolio.allowed_to(User.current, :view_project).any?
```
The check on `.active` is not necessary here as something is only `allowed_to` in an active project. 

Which brings me to an interesting and as of now undiscussed use case: How do we display archived/inactive portfolios and do we allow activating them?

Currently, we do display inactive portfolios (if filtered for them)

<img width="890" height="466" alt="Image" src="https://github.com/user-attachments/assets/ed170cad-bbd7-43c9-b681-ad2984fdcf8f" />

But clicking on the link will lead to a 403. 

For now I would suggest to have a prefix or some other indication to visually convey that the portfolio is inactive and no longer have a link in this case. I would for now disregard the use case of reactivating the portfolio from this page.

EinLama replied:
Created https://community.openproject.org/wp/69114 as a result of our discussion.

ulferts commented on app/components/portfolios/index_sub_header_component.html.erb:
```diff
@@ -0,0 +1,37 @@
+        )
+
+        subheader.with_filter_component do
+          render(Projects::ProjectFilterButtonComponent.new(query: @query, disabled: @disable_buttons))
```
This leads to all the filters also available for projects to be available here as well which is nice for the most part except for the names:

<img width="988" height="412" alt="Image" src="https://github.com/user-attachments/assets/df0110fd-6c36-4587-aa78-1458526ae084" />

I would ignore this for now, though and see if we can later on either find a general name, hide those anyway since some doe not seem applicable or have distinct names per module.

ulferts commented on spec/support/pages/portfolios/index.rb:
```diff
@@ -0,0 +1,172 @@
+      def filter_by_public(value)
+        set_filter("public", "Public", "is", [value])
+        wait_for_reload
+      end
```
Some of those helpers are not actually used and can be removed.

EinLama replied:
When cleaning up, I kept the helpers in that I thought we will need in the near future when overhauling the filters that we want to display on the portfolios page.

Admittedly, this is a case of YAGNI. I will remove some more of these helpers. I will keep the `more` menu helper, since the `sort by` feature is bound to be implemented soon. To make it more useful, I added a spec that checks for the presence of the zen mode.

ulferts reviewed (changes requested):
I am not yet done with the review but I want to document the current state of this PR. The general structure looks good and it is nice to see the module come to life. 
So far I have found a couple to things that would need to be addressed and will continue the review soon.

EinLama committed:
[#68812] Unmask portfolios/new

It was disabled by the double resources statement previously

EinLama committed:
[#68812] Simplify `all_descendants`

There is no need to recursively fetch all child projects. The method
already does that. Also use existing scopes to fetch programs and
portfolios.

EinLama committed:
[#68812] Clean up portfolio page helpers

Add spec for the Zen mode

EinLama committed:
[#68812] Remove unnecessary call to `#active`

Only active projects are considered allowable.

EinLama committed:
[#68812] Permission check for new portfolio button

Also restricts visibility to the descendants that are visible for the
current user.

ulferts commented on app/components/portfolios/details_component.rb:
```diff
@@ -0,0 +1,67 @@
+
+    def all_subprojects
+      all_descendants.project
+    end
```
In this case, using `all_descendants.select { |it| it.project }` and doing the same for programs might be faster as it would result in only one DB query instead of having two. But I didn't measure so I don't know.

EinLama replied:
It depends on the number of descendants. I did a quick benchmark with 100 subprograms and 300 subprojects and relying on the database was 4x quicker than counting in Ruby. With a handful of projects and programs, it was the other way around, which makes sense.

Despite 4x seeming a lot, _both_ options are fast. So I suggest we stick with the database approach to be future-proof 🤔 Counting a few objects in Ruby will always be reasonably quick, but as the collection grows, the DB becomes more efficient.

EinLama committed:
[#68812] Load portfolios with favorited by user

ulferts commented on spec/support/pages/portfolios/index.rb:
```diff
@@ -0,0 +1,145 @@
+      end
+
+      def within_portfolio_list(&)
+        within "#portfolios-index-container", &
```
You could use the test selector here via `within_test_selector` and by that wouldn't need to have an id on that element to begin with.

ulferts commented on app/components/portfolios/index_page_header_component.html.erb:
```diff
@@ -0,0 +1,39 @@
+          content_arguments: { data: { controller: "projects-zen-mode", target: "projects-zen-mode.button", action: "click->projects-zen-mode#performAction" } }
+        ) do |item|
+          item.with_leading_visual_icon(icon: "screen-full")
+        end
```
Clicking on the button currently does nothing.

EinLama replied:
Thank you for noticing this. It worked in the beginning, but I never checked it again. Fixed now ✅

ulferts commented on app/controllers/portfolios/menus_controller.rb:
```diff
@@ -0,0 +1,43 @@
+  class MenusController < ApplicationController
+    # No authorize as every user (or logged in user)
+    # is allowed to see the menu.
+    no_authorization_required! :show
```
This is not critical but I would use the same permission check here as in the portfolios_controller. Would be consistent as the two are actually a pair. And who knows if one or the other more sensitive information finds its way into the menu in the long run.

EinLama replied:
I originally took the easy route and replicated what the other menu controllers were doing. But it does make sense to check the permissions here. Did that 👍

ulferts commented on config/initializers/menus.rb:
```diff
@@ -31,6 +31,18 @@
+            icon: "briefcase",
+            if: ->(_) {
+              OpenProject::FeatureDecisions.portfolio_models_active? &&
+                (User.current.logged? || !Setting.login_required?) &&
```
I think the checks in this line are not needed here since we are already checking for the user to have permissions.

EinLama replied:
If the portfolio is public, AFAIK an anonymous, not-logged-in user has permission to `view_project` that portfolio. So having or not having this line in there **can** make a difference. 🤔

I am not sure about the details of our different configurations, let alone our customers various configurations. So I chose to include this line as it is consistent with other menu points, such as `news` and `work packages`.

EinLama replied:
P.S.: I have not yet looked deeper into `Setting.login_required?` and for what use cases it exists.

ulferts reviewed (changes requested):
Hey @EinLama only found minor things now. Looking good to be merged soon.

EinLama committed:
[#68812] Fix IndexComponentSpec

EinLama committed:
[#68812] Repair Portfolios Zen Mode

EinLama committed:
[#68812] Authorize portfolios menu controller

ulferts reviewed (approved):
Looks good to be merged. The open points are tracked and accepted for now.

Good job on this one, @EinLama .
